src/pages/blog/the-end-of-password-pain-building-frictionless-authentication-at-the-guardian.md

```diff
@@ -0,0 +1,155 @@
+    <figcaption>
+        The create account flow demonstrating account verification with a passcode and setting a password.
+        <i>Illustration: Mahesh Makani</i>
+    </figcaption>
+</figure>
+
+However this introduced a small regression in our security posture and user experience. While new readers going through the create account flow got a passcode for verification, existing readers who accidentally went through the create account flow (ie readers who do not know they already had a Guardian account), were still receiving an email saying they already had an account and links to either reset their password or sign in, causing a mismatched user experience. It also meant we accidentally introduced a way of identifying if a given email address had a Guardian account or not, called “[account enumeration](https://owasp.org/www-project-web-security-testing-guide/v42/4-Web_Application_Security_Testing/03-Identity_Management_Testing/04-Testing_for_Account_Enumeration_and_Guessable_User_Account)”. If a given attacker is able to find out if an account exists or not, this could potentially make it easier for the attacker to use previously stolen or leaked credentials. This can be solved by also giving existing users the same user experience as new users, which would make it difficult to tell if a given account exists or not. The solution to this is discussed later in the “remaining journeys” section.
+
+Adding email verification via passcode to account creation reduced the number of readers failing to verify their account and setting a password from 17% to 11%, which corresponds to ~35% increase in verified readers.
+
+## Password reset
+
+Once passcodes had been implemented for account creation, we moved on to the other main area where links were causing an issue, the password reset journeys. The same problems mentioned there also apply here.
+
+A user will request a password reset when they are unable to remember their password for sign in with email and password. The approach we took for implementing passcodes for password reset was similar to that we used for account creation.
+
+The main consideration this time was to make sure the reader would get the same experience regardless of whether they had an account or not. For readers who have an account, we generate an OTP for password reset and send it to their email, and we show the page to input the code. For readers who don’t have an account, but accidentally end up in this flow, we don’t generate an OTP or create an account, but we still do show the page to input the code, despite not having sent them one. This helps prevent the account enumeration issue mentioned in the previous section. We try to guide readers where possible, for example with specific messaging saying that if a reader does not receive an email, they likely do not have an account, with a call to action to create an account.
```

Comment:
Is it worth mentioning that at some point we might want to send users an email to let them know they don't have an account yet?